### ISSUE TITLE
Unit tests fail with Git 2.38.1 and later

### ISSUE THREAD

**Issue by 9999years** [haskell/cabal#10312]
The `UnitTests.Distribution.Client.VCS` Git tests fail on versions of Git after 2.38. [CVE-2022-39253](https://github.blog/open-source/git/git-security-vulnerabilities-announced/) found vulnerabilities related to local clones of Git repositories:

> In addition to the attack described by CVE-2022-39253, permitting a repository to clone arbitrary repositories on your filesystem via submodules is an unsafe practice. This can be exploited to clone repositories containing sensitive content from well-known locations (such as “/home/username/.git”, which is often used to store sensitive information such as SSH keys).

In Git 2.38.1 and later, these tests will fail with a `fatal: transport 'file' not allowed` message.

```
$ git --version
git version 2.46.0
$ cabal test cabal-install:long-tests --test-options="-p git"
...
fatal: transport 'file' not allowed
fatal: clone of '/private/var/folders/z5/fclwwdms3r1gq4k4p3pkvvc00000gn/T/vcstest-40714/submodules/dyfqwnccfzhmdszwuzpb' into submodule path '/private/var/folders/z5/fclwwdms3r1gq4k4p3pkvvc00000gn/T/vcstest-40714/submodules/vtashfuyukquurrwfupg/file/C' failed
Cloning into '/private/var/folders/z5/fclwwdms3r1gq4k4p3pkvvc00000gn/T/vcstest-40715/src/file/B'...
Long-running tests
  UnitTests.Distribution.Client.VCS
    git
      check VCS test framework:                                                 FAIL (18.58s)
        *** Failed! Exception: 'ExitFailure 128' (after 2 tests and 5 shrinks):
        BranchingRepoRecipe [Left (TaggedCommits "tag_TFDN" [Commit [Right (SubmoduleAdd "file/B" "ejrudwfzgssxxmfwgnfx" (Commit [Left (FileUpdate "file/D" "]PD;[9~2z+")]))]])]
        Use --quickcheck-replay="(SMGen 11397686290392525576 15352236379934602111,10)" to reproduce.
        Use -p '/git.check VCS test framework/' to rerun this test only.
```

We could fix this by adding `-c protocol.file.allow=always` to the relevant Git commands, but that would reintroduce this security issue. I think this would be fine, as long as we only do it in the test suite.

**Comment by geekosaur** [comment]
I believe the assumption is that these tests are only run in CI, which [already works around this](https://github.com/haskell/cabal/blob/master/.github/workflows/validate.yml#L136-L137).

**Comment by 9999years** [comment]
The [CONTRIBUTING.md](https://github.com/haskell/cabal/blob/a092bc8acd9387f63983952d51bc07c37053062d/CONTRIBUTING.md#running-tests) says:

> To run tests locally with `cabal`, you will need to know the name of the test suite you want. Cabal and cabal-install have several.
>
> ...
>
> `cabal-install:integration-tests2` are integration tests on some top-level API functions inside the `cabal-install` source code.

It provides explicit instructions for running tests locally. Either way, we _should_ be able to run tests locally.

**Comment by geekosaur** [comment]
Yes, but there don't seem to be good ways to deal with this on a level above individual tests' repos and the user or global configuration. (I personally have `file` enabled, but that's because of a local repo which is stored in Dropbox.)

**Comment by ulysses4ever** [comment]
For posterity: duplicates #9626.
